File: api_nestjs_distributed_order_platform/src/application/use-case/order/create-order.use-case.ts
import {
    BadRequestException,
    Inject,
    Injectable,
    Logger,
    NotFoundException,
} from "@nestjs/common";
import { InjectConnection } from "@nestjs/sequelize";
import { Sequelize } from "sequelize-typescript";
import type { ClientRepositoryInterface } from "src/application/port/client.repository";
import type { EventPublisherRepositoryInterface } from "src/application/port/event-publisher.repository";
import type { OrderRepositoryInterface } from "src/application/port/order.repository";
import type { ProductRepositoryInterface } from "src/application/port/product.repository";
import type { StockRepositoryInterface } from "src/application/port/stock.repository";
import { OrderItemEntity } from "src/domain/entities/order-item.entity";
import { OrderEntity } from "src/domain/entities/order.entity";
import { StockEntity } from "src/domain/entities/stock.entity";
import { CreateOrderInput } from "src/shared/interfaces/order.interface";
import { CLIENT_REPOSITORY } from "src/shared/tokens_nest/client.token";
import { ORDER_REPOSITORY } from "src/shared/tokens_nest/order.token";
import { PRODUCT_REPOSITORY } from "src/shared/tokens_nest/product.token";
import { EVENT_PUBLISHER } from "src/shared/tokens_nest/rabbitmq.token";
import { STOCK_REPOSITORY } from "src/shared/tokens_nest/stock.token";

@Injectable()
export class CreateOrderUseCase {
    private readonly logger = new Logger(CreateOrderUseCase.name);

    constructor(
        @Inject(ORDER_REPOSITORY)
        private readonly orderRepository: OrderRepositoryInterface,
        @Inject(CLIENT_REPOSITORY)
        private readonly clientRepository: ClientRepositoryInterface,
        @Inject(PRODUCT_REPOSITORY)
        private readonly productRepository: ProductRepositoryInterface,
        @Inject(STOCK_REPOSITORY)
        private readonly stockRepository: StockRepositoryInterface,
        @InjectConnection()
        private readonly sequelize: Sequelize,
        @Inject(EVENT_PUBLISHER)
        private readonly eventPublisher: EventPublisherRepositoryInterface,
    ) {}

    async execute(input: CreateOrderInput): Promise<OrderEntity> {
        if (!input.items?.length) {
            throw new BadRequestException('Order must have at least one item');
        }

        const client = await this.clientRepository.findById(input.clientId);
        if (!client) {
            throw new NotFoundException('Client not found');
        }
        if (!client.canCreateOrder()) {
            throw new BadRequestException('Inactive client cannot create orders');
        }

        const { createdOrder, reservedStocks } = await this.sequelize.transaction(
            async (transaction) => {
                const order = new OrderEntity(input.clientId);
                const reservedStocks: StockEntity[] = [];

                for (const itemInput of input.items) {
                    const product = await this.productRepository.findById(
                        itemInput.productId,
                    );
                    if (!product) {
                        throw new NotFoundException(
                            `Product not found: ${itemInput.productId}`,
                        );
                    }
                    if (!product.canBeOrdered()) {
                        throw new BadRequestException(
                            `Product inactive: ${itemInput.productId}`,
                        );
                    }

                    const stock = await this.stockRepository.reserve(
                        itemInput.productId,
                        itemInput.quantity,
                        transaction,
                    );
                    reservedStocks.push(stock);

                    order.addItem(
                        new OrderItemEntity(
                            itemInput.productId,
                            itemInput.quantity,
                            product.getPrice(),
                        ),
                    );
                }

                const createdOrder = await this.orderRepository.create(order, transaction);

                return { createdOrder, reservedStocks };
            },
        );

        try {
            await this.eventPublisher.publishCreatedOrder(createdOrder);
        } catch (error) {
            this.logger.error(
                `Failed to publish OrderCreated for order ${createdOrder.getId()}`,
                error instanceof Error ? error.stack : String(error),
            );
        }

        for (const stock of reservedStocks) {
            try {
                await this.eventPublisher.publishStockReserved(stock);
            } catch (error) {
                this.logger.error(
                    `Failed to publish StockReserved for product ${stock.getProductId()}`,
                    error instanceof Error ? error.stack : String(error),
                );
            }
        }

        return createdOrder;
    }
}

File: api_java_distributed_order_platform/src/main/java/com/project/order/application/usecase/order/CreateOrderUseCase.java
package com.project.order.application.usecase.order;

import com.project.order.application.exception.BusinessRuleException;
import com.project.order.application.exception.ConflictException;
import com.project.order.application.exception.NotFoundException;
import com.project.order.application.port.ClientRepository;
import com.project.order.application.port.EventPublisherRepository;
import com.project.order.application.port.OrderRepository;
import com.project.order.application.port.ProductRepository;
import com.project.order.application.port.StockRepository;
import com.project.order.domain.model.Client;
import com.project.order.domain.model.Order;
import com.project.order.domain.model.OrderItem;
import com.project.order.domain.model.Product;
import com.project.order.domain.model.Stock;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.dao.OptimisticLockingFailureException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Fluxo: valida cliente -> por item, valida produto + reserva estoque ->
 * cria o pedido. Tudo dentro de UMA transação de banco (@Transactional):
 * se qualquer item falhar (produto inativo, estoque insuficiente, conflito
 * de versão), a transação inteira desfaz as reservas já feitas nos itens
 * anteriores. É o Spring cuidando do que, no NestJS, a gente fazia na mão
 * com sequelize.transaction(...).
 */
@Service
public class CreateOrderUseCase {

    private static final Logger log = LoggerFactory.getLogger(CreateOrderUseCase.class);

    private final OrderRepository orderRepository;
    private final ClientRepository clientRepository;
    private final ProductRepository productRepository;
    private final StockRepository stockRepository;
    private final EventPublisherRepository eventPublisher;

    public CreateOrderUseCase(
            OrderRepository orderRepository,
            ClientRepository clientRepository,
            ProductRepository productRepository,
            StockRepository stockRepository,
            EventPublisherRepository eventPublisher
    ) {
        this.orderRepository = orderRepository;
        this.clientRepository = clientRepository;
        this.productRepository = productRepository;
        this.stockRepository = stockRepository;
        this.eventPublisher = eventPublisher;
    }

    public record OrderItemRequest(String productId, int quantity) {}

    @Transactional
    public Order execute(String clientId, List<OrderItemRequest> items) {
        if (items == null || items.isEmpty()) {
            throw new BusinessRuleException("Order must have at least one item");
        }

        Client client = clientRepository.findById(clientId)
                .orElseThrow(() -> new NotFoundException("Client not found"));
        if (!client.canCreateOrder()) {
            throw new BusinessRuleException("Inactive client cannot create orders");
        }

        Order order = Order.create(clientId);

        for (OrderItemRequest itemRequest : items) {
            Product product = productRepository.findById(itemRequest.productId())
                    .orElseThrow(() -> new NotFoundException("Product not found: " + itemRequest.productId()));
            if (!product.canBeOrdered()) {
                throw new BusinessRuleException("Product inactive: " + itemRequest.productId());
            }

            Stock stock = stockRepository.findById(itemRequest.productId())
                    .orElseThrow(() -> new NotFoundException("Stock not found: " + itemRequest.productId()));

            try {
                stock.reserve(itemRequest.quantity());
            } catch (IllegalArgumentException e) {
                throw new BusinessRuleException(e.getMessage());
            }

            try {
                stockRepository.update(stock);
            } catch (OptimisticLockingFailureException e) {
                throw new ConflictException("Stock was modified by another request, retry");
            }

            order.addItem(new OrderItem(itemRequest.productId(), itemRequest.quantity(), product.getPrice()));
            publishQuietly(() -> eventPublisher.publishStockReserved(stock), "StockReserved", stock.getProductId());
        }

        Order created = orderRepository.create(order);
        publishQuietly(() -> eventPublisher.publishCreatedOrder(created), "OrderCreated", created.getId());

        return created;
    }

    private void publishQuietly(Runnable action, String eventName, String subjectId) {
        try {
            action.run();
        } catch (Exception e) {
            log.error("Failed to publish {} for {}", eventName, subjectId, e);
        }
    }
}
